x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/accesscontrol/IndicesAccessControl.java

```diff
@@ -64,12 +71,27 @@ public boolean isGranted() {
         return granted;
     }
 
-    public Collection<?> getDeniedIndices() {
-        return this.indexPermissions.entrySet()
+    /**
+     * @return A collection on index names for which the current request was denied access, divide in regular and
+     * {@link #isRestrictedIndex(String) restricted} indices.
+     */
+    public DeniedIndices getDeniedIndices() {
+        final Map<Boolean, Set<String>> deniedIndices = this.indexPermissions.entrySet()
             .stream()
             .filter(e -> e.getValue().granted == false)
             .map(Map.Entry::getKey)
-            .collect(Collectors.toUnmodifiableSet());
+            .collect(Collectors.groupingBy(this::isRestrictedIndex, Collectors.toUnmodifiableSet()));
+        final Set<String> regularIndices = deniedIndices.getOrDefault(Boolean.FALSE, Set.of());
+        final Set<String> restrictedIndices = deniedIndices.getOrDefault(Boolean.TRUE, Set.of());
+        return new DeniedIndices(regularIndices, restrictedIndices);
+    }
+
+    /**
+     * @return {code true} if the provided index {@code name} is a
+     * {@link org.elasticsearch.xpack.core.security.index.RestrictedIndicesNames restricted index}
+     */
+    public boolean isRestrictedIndex(String name) {
+        return restrictedIndices.test(name);
     }
```

Comment:
This method is weird because it is not always consistent and the answer may not be correct. Whether an index is restricted has a definitive answer in a specific version. But that is not the case with this method because sometimes the `restrictedIndices` predicate always returns `false`. I'd be less concerned if this method is `private`. But right now it seems rather confusing.